### ISSUE TITLE
Consider replacing astropy's progress bar machinery with tqdm

### ISSUE THREAD

**Issue by eteq** [astropy/astropy#4738]
[tqdm](https://pypi.python.org/pypi/tqdm) is a package that implements easy/pretty progress bars in python.  While we have something like this in `astropy.utils.console`, it might be wise to switch to tqdm in the interest of better code sharing/reuse.

One concern is how hard it will be to adapt our machinery to using this.  If the interface is too different it's probably not worth it.

Another concern that needs checking is whether or not the notebook progress bar in tqdm is up-to-snuff (because the astropy one works quite well).

Originally suggested by @adrn

**Comment by pllim** [comment]
I would like to offer a different viewpoint -- "If it ain't broken, don't fix it." Is it worth adding yet another dependency just for a prettier progress bar?

**Comment by adrn** [comment]
The reason is not that it's prettier, but rather that it doesn't really make sense for astropy to maintain a progress bar utility. For the sake of the open-source community and the modularity of astropy, I think it would make sense to switch. Of course, this should be weighed with the practicality of switching...

**Comment by astrofrog** [comment]
Just to be clear, would we bundle it or have it as a dependency?

**Comment by adrn** [comment]
@astrofrog I would say dependency. If `tqdm` is not installed, I think the user could live without a progress bar? The Astropy `ProgressBar` class is only used in [astropy/utils/data.py](https://github.com/astropy/astropy/blob/master/astropy/utils/data.py) and [astropy/utils/data.py](https://github.com/astropy/astropy/blob/master/astropy/io/votable/validator/main.py)

**Comment by pllim** [comment]
To say it is "*only used*" is misleading. Whenever remote data is downloaded, you will see that progress bar. I still think it's a bit excessive to require an extra dependency for a low-level util like this, but that's just me.

**Comment by mhvk** [comment]
I think it is a good idea to swap, since really we should not maintain that kind of code, but on the other hand I do think the default should be to have a progress bar without having to depend on the user installing it. Since`tqdm` is not exactly large, (indeed, it seems fairly similar to our current `console.py`), could we not just bundle it?

**Comment by cdeil** [comment]
I wanted to use the Astropy progress update utils today for the first time.
The use case is to report to users progress of HiPS tile fetching from the internet and tile drawing to make sky images: https://github.com/hipspy/hips/issues/80

So I looked at the examples here:
* http://docs.astropy.org/en/stable/api/astropy.utils.console.Spinner.html
* http://docs.astropy.org/en/stable/api/astropy.utils.console.ProgressBar.html

Some observations / feedback as a naive first-time user:
* None of the examples given is complete that I can just copy & paste them and try them out.
* The spinner example calls `next` on a generator, which doesn't work on Python 3 where it's `__next__`, this needs to be fixed.
* Why is it `bar.update()` and `s.next()`? This difference is unnecessary and hard to remember, no?
* By default in the Jupyter notebook there's no output at all from the progress bar. Wouldn't it be possible / better default behaviour to show text output like on the terminal? I'm not sure how to use the progress bar from my `hips` package so that it shows nice progress for users both from the console and Jupyter (and some progress even in cases where they don't have the Jupyter widgets enabled like I have at the moment).

This is the full example that I used in the end to try it, we could put that in the docstring:
```
In [3]: %paste
from astropy.utils.console import ProgressBar
import time

items = range(10)
for item in ProgressBar(items):
    # Do some computation with `item` here
    time.sleep(0.2)

## -- End pasted text --
|================================================================================================================|  10 / 10  (100.00%)         2s
```

The discussion here has stalled over a year ago. I can see how this is a tough decision. Is there a way to resolve the discussion here and make a decision what to do?

Some of the comments above lean towards switching out the implementation and (if I understand correctly), change the implementation of the existing classes to be wrappers around `tqdm` and provide backward compatibility for users of the `astropy.utils`. Is it possible? Is it desirable given the point I made about the spinner and progress bar API above or would complete deprecation of this part of `astropy.utils` be better (I mean via the docs, i.e. just say that external packages like `hips` should consider using something like `tqdm` directly)?

If I get some feedback here, I could make a PR fixing / improving the docstrings of `Spinner` and `ProgressBar` a bit, but I don't have time to e.g. try switching to `tqdm` (and I'm not familiar with the implementations at all).

**Comment by Cadair** [comment]
I think in general this is a good idea. I mean the "simplest" solution is to change all internal uses of progress bar to tqdm and then just deprecate the existing implementation?

**Comment by MridulS** [comment]
Just wanted to bring this up again after 8 years :)
[tqdm](https://github.com/tqdm/tqdm/) is pretty much used everywhere now, and I am not sure how actively is the ProgressBar inside astropy is maintained. I tried switching to tqdm in astroquery and it shouldn't be that big of difference for downstream users (IMO).

With astropy's progressbar:
<img width="1172" alt="Screenshot 2024-02-20 at 10 36 54 PM" src="https://github.com/astropy/astropy/assets/5363860/cf28a0eb-c3d5-4f80-82cd-a5487fe9e73b">

With tqdm:
<img width="1204" alt="Screenshot 2024-02-20 at 10 38 41 PM" src="https://github.com/astropy/astropy/assets/5363860/560396d5-772b-4e6b-a6ce-15285bd49d38">

Let me know if there is interest in pursuing it :)

**Comment by pllim** [comment]
The more controversial discussion is adding it as a required dependency. We can't even agree on Scipy... 

* https://github.com/astropy/astropy/issues/12448

And even older but broader discussions:

* https://github.com/astropy/astropy/issues/10342

**Comment by tactipus** [comment]
we should make it a dependency. we should it duke it out to see which side prevails :3

**Comment by WilliamJamieson** [comment]
The main opposition to `scipy` surrounds how "big" its install is and how complicated it can be to build from source. `tqdm` appears to be ~78kB and entirely Python code with no serious dependencies itself. So I think this is an entirely reasonable thing to do. Especially if @MridulS is correct in it being a fairly simple change.

**Comment by Cadair** [comment]
tqdm is already in the Debian archive, I used it from my shell the other day 😁

**Comment by WilliamJamieson** [comment]
Here it is: https://packages.debian.org/buster/python3-tqdm

**Comment by olebole** [comment]
I can confirm that tqdm is not a problem for me. I'd support this PR; a progress bar is nothing native to Astropy and therefore should be replaced by something external if possible.

**Comment by bmorris3** [comment]
My two cents: tqdm works great, astropy doesn't need to maintain an alternative. Bin it. 🗑️

**Comment by tactipus** [comment]
i've used tqdm for a small project. it is a PDF scanner that uses tdqm to tell you the progress on the scan. it was QUITE painless to incorporate compared to the scanning libraries.

**Comment by taldcroft** [comment]
I use `tqdm` regularly and never use the astropy progress bar. Bin it!

**Comment by tactipus** [comment]
who gave me a thumbs up…?

**Comment by DBerke** [comment]
My 2¢: I've used `tqdm` quite a bit and it's quite solid; it's both very easy to use the default by just wrapping any iterable, but offers a ton of depth and customizability if you want/need to really tinker with it.

**Comment by matteobachetti** [comment]
I didn't know that there was an Astropy progress bar, and I use tqdm all the time 😅
If the move is painless enough, I'd go for it. If tqdm becomes stale, it would be easy to bundle or repackage anyway

**Comment by mhvk** [comment]
Also in favour. On the implementation side, a progress bar feels rather optional (hardly needed for CI!), so I'd put it in `[all]`

**Comment by adonath** [comment]
Same here. In Gammapy we have been using `tqdm` to show progress for downloads and data reduction since >5 years. It was easy to implement and we never had any installation / dependency issues. However we have it as an optional dependency, so if it is not installed no progress bar is shown. We have achieved this with a simple mockup for `tqdm`: https://github.com/gammapy/gammapy/blob/main/gammapy/utils/pbar.py#L7

**Comment by pllim** [comment]
@MridulS , based on the overwhelming support for the switch, are you interested in a draft PR to explore this?

**Comment by MridulS** [comment]
Will send one soon! right now fighting random seeds :p

**Comment by pllim** [comment]
<img src="https://m.media-amazon.com/images/I/81ErNcODNBL._AC_UF1000,1000_QL80_.jpg" height="400"/>

**Comment by astrofrog** [comment]
We do use a progress bar in astropy so I think tqdm should be a required dependency?

**Comment by mhvk** [comment]
I don't think it should be required - it is not essential for anything and (I presume) turned off by default in CI. It just means that the default for showing the progress bar should be "use if installed". With the IERS data now a required dependency, a typical astropy user will no longer see the progress bar at all.

**Comment by pllim** [comment]
> I don't think it should be required

I think the impact would be annoying downstream since `download_file` is widely used. A downstream package would find itself suddenly with a "broken" progress bar unless they add the "optional" `tqdm` into their own dependency even though they are using `astropy.utils.data.download_file` from us.

**Comment by pllim** [comment]
> how did we make the current progress bar?

We rolled our own, see

https://github.com/astropy/astropy/blob/2597e904f84362c99c8b0ef948ef69bd23029830/astropy/utils/console.py#L445

**Comment by mhvk** [comment]
I dislike depending on trivial additions, but since we shipped something ourselves (which was not that much smaller), I guess it is OK. I just wish it was easy with `pip` to have a default choice for dependencies that was not the minimal set (but, say, the `[recommended]` one...).

**Comment by MridulS** [comment]
Alright so I went in to start changing the `ProgressBar` plumbing to tqdm but it looks like astropy's `ProgressBar` is not just a progress bar. It also has logic for running a function with multiprocessing!! (https://github.com/astropy/astropy/blob/2597e904f84362c99c8b0ef948ef69bd23029830/astropy/utils/console.py#L709) which is used in a couple of places.

If we do deprecate `ProgressBar` we may have to keep this bit of code around.

**Comment by pllim** [comment]
> which is used in a couple of places

Hmm... Where exactly?

**Comment by MridulS** [comment]
> Hmm... Where exactly?

https://github.com/astropy/astropy/blob/884bcacb79d4903175bedfc40706057860543e59/astropy/utils/data.py#L1749-L1759

https://github.com/astropy/astropy/blob/884bcacb79d4903175bedfc40706057860543e59/astropy/io/votable/validator/main.py#L136

**Comment by pllim** [comment]
Right, the validator. Yeah, I remember now. It was designed to download multiple sources at once for validation. If PyVO eventually takes it over, maybe it is not such a big deal for `astropy`.

https://astroconda.org/aux/vo_databases/daily_html/

*But* the one in `download_files_in_parallel` would be tough to replace, since `astropy.utils.data` stuff is widely used downstream.

Does `tqdm` not support multiprocessing at all? That is a shame.

Should we just close this as won't fix then?

**Comment by MridulS** [comment]
> Does tqdm not support multiprocessing at all? That is a shame.

It does but it is just a progress bar, you would still need to write the actual code to do the multiprocessing bits. Astropy's porgressbar isn't really named properly, it's doing a lot more than just a progress bar 😅 

> Should we just close this as won't fix then?

Yeah I think we can just close this issue, let's keep it https://github.com/astropy/astropy/issues/4738#issuecomment-202409768 till it starts annoying us.

**Comment by astrofrog** [comment]
> It does but it is just a progress bar, you would still need to write the actual code to do the multiprocessing bits. Astropy's porgressbar isn't really named properly, it's doing a lot more than just a progress bar 😅

I've used tqdm with multiprocessing - and for simple use cases it's been as easy as just setting up a multiprocessing.Pool so it might not be too much work.

**Comment by pllim** [comment]
@astrofrog , what are you suggesting here in terms of actionable item?

**Comment by saimn** [comment]
Since both cases are dealing with parallel download, it would probably be better to use a thread pool (e,g, `concurrent.futures`) instead of a process one. (Or some package dedicated to parallel download, e.g. https://github.com/Cadair/parfive but maybe overkill for just those 2 cases)

**Comment by pllim** [comment]
I would be more comfortable if parfive isn't in Stuart's personal account and has recent releases...

If we are serious about parfive, we are talking about requiring also `tqdm` *and* `aiohttp`.

https://github.com/Cadair/parfive/blob/3b049c55f5cd35a607aafdfecfbccabf7bc71e29/setup.cfg#L19-L21

**Comment by tactipus** [comment]
I'm down to code whatever needs to be done.

**Comment by saimn** [comment]
@pllim - not really serious, for 2 simple cases using directly a thread pool would be simpler.

**Comment by pllim** [comment]
I still don't see much point in introducing tqdm if we cannot get rid of anything in the end. Progress bar tech isn't going to change much and it is currently working.

**Comment by astrofrog** [comment]
At the very least we could always consider just recommending to users in the docs to consider using tqdm instead.

**Comment by Cadair** [comment]
> I would be more comfortable if parfive isn't in Stuart's personal account and has recent releases...

*ouch* :stuck_out_tongue: 

It got a bit of love recently, it's actually reasonably active for such a project which does what it intends to.

> If we are serious about parfive, we are talking about requiring also `tqdm` _and_ `aiohttp`.

I am considering [ditching aiohttp](https://github.com/Cadair/parfive/issues/143) in favour of httpx as it's a little lighter weight (aiohttp is also a full http server library). If that's enough to make a difference then I would bump it up the list .
